src/impls.rs

```diff
@@ -1,6 +1,52 @@
 use crate::encoding;
 use crate::error::Result;
 
+macro_rules! derrive_base_encoding {
+    (#[$doc:meta] $type:ident, $encoding:expr;) => {
+
+        #[$doc]
+        #[derive(PartialEq, Eq, Clone, Copy, Debug)]
+        pub(crate) struct $type;
+
+        impl BaseCodec for $type {
+            fn encode<I: AsRef<[u8]>>(input: I) -> String {
+                $encoding.encode(input.as_ref())
+            }
+
+            fn decode<I: AsRef<str>>(input: I) -> Result<Vec<u8>> {
+                Ok($encoding.decode(input.as_ref().as_bytes())?)
+            }
+        }
+    };
+    (#[$doc:meta] $type:ident, $encoding:expr; $(#[$doc2:meta] $type2:ident, $encoding2:expr;)+) => {
+        derrive_base_encoding! (#[$doc] $type, $encoding;);
+        derrive_base_encoding! ($(#[$doc2] $type2, $encoding2;)+);
+    };
```

Comment:
That's not needed. You can change the rule at the top.

Wrap it in a `($(…)*)`. Then wrap the whole contents (the stuff between the curly brackets) in `$(…)*`.

So it reads line:

```rust
($(#[$doc:meta] $type:ident, $encoding:expr;)*) => {
    $(
…
    )*
};
```

I also had to look that up. https://medium.com/@phoomparin/a-beginners-guide-to-rust-macros-5c75594498f1 helped a lot.